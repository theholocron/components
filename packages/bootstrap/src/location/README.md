# Location Component

A React Provider component for storing the users current location.

## Usage

```typescript
import { Location } from "@theholocron/bootstrap";
import * as React from "react";

interface AppProps {
    Component: React.ReactNode;
    pageProps?: any;
}

export default function App (props: AppProps) {
    const { Component, pageProps } = props;

    return (
        <Location.Provider>
            <Component {...pageProps} />
        </Location.Provider>;
    );
}
```

### Hook

There is a hook that comes along with the library for detecting the current location.

```typescript
import { useLocation } from "@theholocron/bootstrap";

function Component () {
    const location = useLocation();

    return <h3>Current location: <code>{location.latitude}</code> (latitude), <code>{location.longitude}</code> (longitude)</h3>;
}
```
