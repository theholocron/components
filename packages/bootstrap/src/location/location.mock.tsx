import { useLocation } from "./index";

export function TestComponent() {
	const { location, isLoading } = useLocation();

	if (isLoading) return <div data-testid="loading">Loading...</div>;

	if (location) {
		return (
			<div data-testid="location">
				<p>Latitude: {location.latitude}</p>
				<p>Longitude: {location.longitude}</p>
				<p>Accuracy: {location.accuracy} meters</p>
			</div>
		);
	}

	return <div data-testid="no-location">No location available</div>;
}
