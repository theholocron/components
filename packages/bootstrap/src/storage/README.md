# Browser Storage

A React Provider component for storing data within browser storage utilizing `window.sessionStorage`.

## Usage

```typescript
import { Storage } from "@theholocron/bootstrap";
import * as React from "react";

const conf: ApplicationConf = {
    application: { id: "mock-app-id" },
};

interface AppProps {
    Component: React.ReactNode;
    pageProps?: any;
}

export default function App (props: AppProps) {
    const { Component, pageProps } = props;

    return (
        <Storage.Provider conf={conf.application}>
            <Component {...pageProps} />
        </Storage.Provider>;
    );
}
```

### `useStorage`

Use this function for getting, setting and removing items from the storage, i.e. `sessionStorage`.

```typescript
import { useStorage } from "@theholocron/bootstrap";

const storage = useStorage();

// send an item to the store
storage.sendTo("mockKey", item);

// grab an item from the store
const item = storage.getFrom("mockKey");

// delete it if you want
storage.removeFrom("mockKey");
```
