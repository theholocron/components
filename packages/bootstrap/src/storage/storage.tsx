import { storage, type TSessionStorage } from "@theholocron/utils-storage";
import * as React from "react";
import { type ApplicationConf } from "../configuration/index";
import type { WithChildren } from "../type";

const Context = React.createContext<TSessionStorage | undefined>(undefined);
Context.displayName = "@theholocron/bootstrap/Storage";

export interface StorageProps extends WithChildren {
	conf: ApplicationConf;
}

export function Provider(props: StorageProps) {
	const value = React.useMemo(() => {
		const vault = storage.session.create();

		vault.registerApp(props.conf.id);
		vault.sendTo("application.name", props.conf.name);
		vault.sendTo("build.version", props.conf?.version ?? "0.0.0");

		return vault;
	}, [props.conf.id, props.conf.name, props.conf.version]);

	return <Context.Provider value={value}>{props.children}</Context.Provider>;
}
Provider.displayName = "@theholocron/bootstrap/Storage";

export function useStorage(): TSessionStorage {
	const context = React.useContext(Context);

	if (context === undefined) {
		throw new Error("useStorage must be used within Storage.Provider!");
	}

	return context;
}
