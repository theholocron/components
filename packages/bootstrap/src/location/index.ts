import { Provider } from "./location.tsx";

export { useLocation } from "./location.tsx";
export type { TLocation } from "./location.tsx";

export default { Provider };
