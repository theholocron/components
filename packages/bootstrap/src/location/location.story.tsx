import { expect, within, userEvent } from "@storybook/test";
import * as React from "react";
import Storage, { useStorage } from "../storage";
import { Default as StorageDefault } from "../storage/storage.story";
import Location, { useLocation } from "./index";
import { TestComponent } from "./location.mock";

export default {
	component: Location.Provider,
	title: "Location",
	decorators: [
        (Story) => (
            <Storage.Provider {...StorageDefault.args}>
                <Story />
			</Storage.Provider>
        ),
    ],
};

// 1. Story: Successful Location Fetch
export const Default = () => (
	<Location.Provider>
		<TestComponent />
	</Location.Provider>
);

// Attempt to use the hook without provider
export const Error = () => {
	const Component = () => {
		try {
			useLocation();
		} catch (error) {
			return <div data-testid="error-message">{error.message}</div>;
		}
		return null; // Render nothing if no error
	};

	return <Component />;
};
Error.play = async ({ canvasElement }) => {
	const canvas = within(canvasElement);

	const errorMessage = await canvas.findByTestId("error-message");
	expect(errorMessage).toHaveTextContent("useLocation must be used within a Location.Provider!");
};

// Utility function for clearing storage in tests
const clearStorage = () => {
	const storage = useStorage();
	storage.clear();
};
// 2. Story: Permission Denied
export const PermissionDenied = () => (
	<Location.Provider>
		<TestComponent />
	</Location.Provider>
);

// 3. Story: Cached Location Used
export const CachedLocation = () => (
	<Location.Provider>
		<TestComponent />
	</Location.Provider>
);

// 4. Story: Error Fetching Location
export const LocationError = () => (
	<Location.Provider>
		<TestComponent />
	</Location.Provider>
);
