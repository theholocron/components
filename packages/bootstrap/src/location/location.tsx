import {
	location,
	type IGeolocationCoordinates,
} from "@theholocron/utils-location";
import * as React from "react";
import { useStorage } from "../storage/";
import type { WithChildren } from "../type";

export type TLocation = {
	location: IGeolocationCoordinates | null;
	isLoading: boolean;
};

type CachedLocationData = {
	coords: IGeolocationCoordinates;
	timestamp: number;
};

const Context = React.createContext<TLocation | undefined>(undefined);
Context.displayName = "@theholocron/bootstrap/Location";

export function Provider(props: WithChildren) {
	const [isLoading, setIsLoading] = React.useState<boolean>(true);
	const [userLocation, setUserLocation] =
		React.useState<IGeolocationCoordinates | null>(null);
	const storage = useStorage();

	React.useEffect(() => {
		const storageKey = "user.location";
		const maximumAge = 1000 * 60 * 15; // 15 minutes
		const now = Date.now();

		const fetchLocation = async () => {
			const cachedData = storage.getFrom<CachedLocationData>(storageKey);
			if (cachedData && now - cachedData.timestamp < maximumAge) {
				setUserLocation(cachedData.coords);
				setIsLoading(false);
				return;
			}

			try {
				const coords = await location.getCurrent();
				setUserLocation(coords);
				storage.sendTo(storageKey, {
					coords,
					timestamp: now,
				});
			} catch (error) {
				console.error("Failed to fetch location:", error);
			} finally {
				setIsLoading(false);
			}
		};

		fetchLocation();
	}, [storage]);

	const value = React.useMemo(
		() => ({ location: userLocation, isLoading }),
		[userLocation, isLoading]
	);

	return <Context.Provider value={value}>{props.children}</Context.Provider>;
}
Provider.displayName = "@theholocron/bootstrap/Location";

export function useLocation(): TLocation {
	const context = React.useContext(Context);

	if (context === undefined) {
		throw new Error("useLocation must be used within a Location.Provider!");
	}

	return context;
}
