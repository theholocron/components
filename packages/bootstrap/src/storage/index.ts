import { type TSessionStorage } from "@theholocron/utils-storage";
import { Provider } from "./storage";

export { useStorage } from "./storage";
export type { TSessionStorage } from "./storage";

export default { Provider };
