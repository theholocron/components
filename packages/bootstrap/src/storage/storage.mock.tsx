import * as React from "react";
import { useStorage } from "./storage";

export function TestComponent () {
	const [isSending, setIsSending] = React.useState(true);
	const [data, setData] = React.useState({});
	const storage = useStorage();

	const handleClick = () => {
		if (isSending) {
			storage.sendTo("test", "value");
		} else {
			storage.removeFrom("test");
		}
		setData(storage.getAll());
		setIsSending(!isSending);
	};

	React.useEffect(() => {
		setData(JSON.parse(window.sessionStorage["@theholocron"]));
	}, []);

	return (
		<div data-testid="mock-storage">
			<button
				onClick={handleClick}
				type="button"
			>
				{isSending ? "Send to Storage" : "Remove from Storage"}
			</button>
			<pre>{JSON.stringify(data, null, 4)}</pre>
		</div>
	);
}
