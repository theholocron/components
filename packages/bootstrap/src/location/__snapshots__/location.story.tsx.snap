// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Location CachedLocation smoke-test 1`] = `
<div data-testid="location">
  <p>
    Latitude: 0
  </p>
  <p>
    Longitude: 0
  </p>
  <p>
    Accuracy: 0 meters
  </p>
</div>
`;

exports[`Location Default smoke-test 1`] = `
<div data-testid="location">
  <p>
    Latitude: 0
  </p>
  <p>
    Longitude: 0
  </p>
  <p>
    Accuracy: 0 meters
  </p>
</div>
`;

exports[`Location Error play-test 1`] = `
<div data-testid="error-message">
  useLocation must be used within a Location.Provider!
</div>
`;

exports[`Location LocationError smoke-test 1`] = `
<div data-testid="location">
  <p>
    Latitude: 0
  </p>
  <p>
    Longitude: 0
  </p>
  <p>
    Accuracy: 0 meters
  </p>
</div>
`;

exports[`Location PermissionDenied smoke-test 1`] = `
<div data-testid="location">
  <p>
    Latitude: 0
  </p>
  <p>
    Longitude: 0
  </p>
  <p>
    Accuracy: 0 meters
  </p>
</div>
`;
